File: cpp/lib/testcontainers/core/ContainerPort.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/core/ContainerPort.hpp"

#include "details/BoxHelper.hpp"

namespace testcontainers {

ContainerPort::ContainerPort(RsContainerPort *rimpl) noexcept
    : rimpl_(rimpl,
             [](RsContainerPort *p) { ::rs_container_port_destroy(details::box_from_raw(p)); }) {}

ContainerPort::ContainerPort(ContainerPort &&other) noexcept = default;

ContainerPort &ContainerPort::operator=(ContainerPort &&other) noexcept = default;

ContainerPort::~ContainerPort() noexcept = default;

ContainerPort ContainerPort::Tcp(std::uint16_t port) noexcept {
  return ContainerPort(::rs_container_port_tcp(port).into_raw());
}

ContainerPort ContainerPort::Udp(uint16_t port) noexcept {
  return ContainerPort(::rs_container_port_udp(port).into_raw());
}

ContainerPort ContainerPort::Sctp(uint16_t port) noexcept {
  return ContainerPort(::rs_container_port_sctp(port).into_raw());
}

bool ContainerPort::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

std::uint16_t ContainerPort::as_u16() const noexcept {
  return rimpl_->rs_container_port_as_u16();
}

bool ContainerPort::is_tcp() const noexcept { return rimpl_->rs_container_port_is_tcp(); }
bool ContainerPort::is_udp() const noexcept { return rimpl_->rs_container_port_is_udp(); }
bool ContainerPort::is_sctp() const noexcept { return rimpl_->rs_container_port_is_sctp(); }

} // namespace testcontainers

File: rust/src/core/container_port.rs
use testcontainers::core::ContainerPort;

pub struct RsContainerPort {
	port: ContainerPort,
}

pub fn rs_container_port_tcp(port: u16) -> Box<RsContainerPort> {
	Box::new(RsContainerPort::new(ContainerPort::Tcp(port)))
}

pub fn rs_container_port_udp(port: u16) -> Box<RsContainerPort> {
	Box::new(RsContainerPort::new(ContainerPort::Udp(port)))
}

pub fn rs_container_port_sctp(port: u16) -> Box<RsContainerPort> {
	Box::new(RsContainerPort::new(ContainerPort::Sctp(port)))
}

pub fn rs_container_port_destroy(port: Box<RsContainerPort>) {
	drop(port);
}

pub fn rs_container_port_vec_pop(vec: &mut Vec<RsContainerPort>) -> Result<Box<RsContainerPort>, String> {
	vec.pop()
		.map(Box::new)
		.ok_or_else(|| "Vector is empty".to_string())
}

impl RsContainerPort {
	pub fn new(port: ContainerPort) -> Self {
		Self { port }
	}

	pub fn rs_container_port_as_u16(&self) -> u16 {
		match self.port {
			ContainerPort::Tcp(port) => port,
			ContainerPort::Udp(port) => port,
			ContainerPort::Sctp(port) => port,
		}
	}

	pub fn rs_container_port_is_tcp(&self) -> bool {
		matches!(self.port, ContainerPort::Tcp(_))
	}

	pub fn rs_container_port_is_udp(&self) -> bool {
		matches!(self.port, ContainerPort::Udp(_))
	}

	pub fn rs_container_port_is_sctp(&self) -> bool {
		matches!(self.port, ContainerPort::Sctp(_))
	}
}

impl Into<ContainerPort> for RsContainerPort {
	fn into(self) -> ContainerPort {
		self.port
	}
}

impl From<ContainerPort> for RsContainerPort {
	fn from(port: ContainerPort) -> Self {
		Self::new(port)
	}
}
